### PR TITLE
add file_operator tools

### PR DESCRIPTION
**Pull request by longsihua2003**
Add file operator like langchain file operator.

## Summary by CodeRabbit

- **New Features**
    - Added classes for file manipulation tools such as writing, appending, reading, deleting, listing directories, copying, and moving files.
    - Introduced a tutorial (`file_tool_usage.ipynb`) demonstrating how to utilize the new file operation tools directly or through an agent.
- **Tests**
    - Included test cases to validate the functionality of the file manipulation tools provided by the `FileToolKit`.

### CODE REVIEW

**Comment by Undertone0809** [comment]
@longsihua2003 plz run `make formatting`

**Comment by Undertone0809** [comment]
You can also add a notebook show how to use it in `/example`

File: promptulate/tools/file/toolkit.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    "AppendFileTool":AppendFileTool,
+    "ReadFileTool":ReadFileTool,
+    "DeleteFileTool":DeleteFileTool,
+    "ListDirectoryTool":ListDirectoryTool,
+    "CopyFileTool":CopyFileTool,
+    "MoveFileTool":MoveFileTool,
+}
+
+
+class FileToolKit(BaseToolKit):
```
Comment (Undertone0809) [inline promptulate/tools/file/toolkit.py:36]: Maybe you can use a simpler definition like python file operation.

```python
file_tools = FileToolKit(root_dir="./xxx",modes=["read", "write", "append'])
agent = ToolAgent(tools=file_tools)
```

**Review by Undertone0809 — APPROVED** [review]

File: promptulate/tools/file/tools.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    description: str = (
+        "Write to a file"
+        "Useful when you need to edit/create and edit a file."
+        "It can edit the specified file in the specified directory/local directory"
+        "If the file does not exist, edit it after it is created."
+        "The output parameters must be:"
+        "file_name(file name), text(related information)."
+    )
+
+    def __init__(self, root_dir: str) -> None:
```
Comment (Undertone0809) [inline promptulate/tools/file/tools.py:18]: All your tool need to add `super().__init__(*args, **kwargs)`

**Review by Undertone0809 — CHANGES_REQUESTED** [review]

File: tests/tools/test_file_tools.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    DeleteFileTool,
+    ListDirectoryTool,
+    MoveFileTool,
+    ReadFileTool,
+    WriteFileTool,
+)
+
+
+class TestFileToolKit:
+    # 创建临时目录
```
Comment (Undertone0809) [inline tests/tools/test_file_tools.py:19]: English is better

**Review by Undertone0809 — CHANGES_REQUESTED** [review]

File: promptulate/tools/file/toolkit.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        selected_tools: The selected tools of the file tool.
+
+    Returns:
+        The instance object of the corresponding tool
+    """
+
+    def __init__(self, root_dir: str = None, modes: Optional[List[str]] = None) -> None:
+        """validate_root_dir"""
+        if root_dir is None or root_dir == "":
+            root_dir = os.getcwd()
```
Comment (Undertone0809) [inline promptulate/tools/file/toolkit.py:50]: `root_dir = root_dir or os.getcwd()` is better

**Review by Undertone0809 — CHANGES_REQUESTED** [review]

File: promptulate/tools/file/toolkit.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    Args:
+        root_dir: The root directory of the file tool.
+        selected_tools: The selected tools of the file tool.
+
+    Returns:
+        The instance object of the corresponding tool
+    """
+
+    def __init__(self, root_dir: str = None, modes: Optional[List[str]] = None) -> None:
```
Comment (Undertone0809) [inline promptulate/tools/file/toolkit.py:47]: This is my implementation, as a reference for your optimization.

```python
class FileToolKit(BaseToolKit):
    def __init__(self, root_dir: str = None, modes: Optional[List[str]] = None) -> None:
        self.root_dir = root_dir
        self.modes = modes or []
        
        for mode in self.modes:
            if mode not in TOOL_MAPPER.keys():
                raise ValueError(
                    f"{tool} does not exist.\n"
                    f"Please select from {list(DefaultTools_str.keys())}"
                )
    
    def get_tools(self) -> List[Tool]:
        return [self.modes[mode] for mode in self.modes]
```

**Review by Undertone0809 — CHANGES_REQUESTED** [review]

File: promptulate/tools/file/tools.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+    def __init__(self, root_dir: str, *args, **kwargs) -> None:
+        self.root_dir = root_dir
+        super().__init__(*args, **kwargs)
+
+    def _run(self, file_name: str) -> str:
+        """Read a file
+
+        Args:
+            query: file_name.
```
Comment (Undertone0809) [inline promptulate/tools/file/tools.py:97]: useless parameter

File: promptulate/tools/file/tools.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+    def __init__(self, root_dir: str, *args, **kwargs) -> None:
+        self.root_dir = root_dir
+        super().__init__(*args, **kwargs)
+
+    def _run(self, file_name: str, text: str) -> str:
+        """Append to a file
+
+        Args:
+            query: file_name and text.
```
Comment (Undertone0809) [inline promptulate/tools/file/tools.py:63]: useless parameter

File: promptulate/tools/file/tools.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+    def __init__(self, root_dir: str, *args, **kwargs) -> None:
+        self.root_dir = root_dir
+        super().__init__(*args, **kwargs)
+
+    def _run(self, file_name: str) -> str:
+        """Delete a file
+
+        Args:
+            query: file_name.
```
Comment (Undertone0809) [inline promptulate/tools/file/tools.py:136]: useless parameter

File: promptulate/tools/file/tools.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+            with open(f"{self.root_dir}/{file_name}", "a") as file:
+                file.write(text)
+        except IOError:
+            return "An error occurred while trying to write to the file."
+        except PermissionError:
+            return "Permission denied: You don't have permission to write to this file."
+        return "Append file successfully"
+
+
+class ReadFileTool(Tool):
```
Comment (Undertone0809) [inline promptulate/tools/file/tools.py:80]: Here is a better pattern if you want to declare parameter to LLM.

```python
class ReadFileTool(Tool):
    class ReadFileToolParameters(BaseModel):
        file_name: str = Field(
            ..., description="The name of the file you want to read."
        )

    name: str = "read-file"
    description: str = (
        "Read a file"
        "Useful when you need to read the content of specified file."
    )
    parameters: BaseModel = ReadFileToolParameters

```

**Review by Undertone0809 — CHANGES_REQUESTED** [review]

File: promptulate/tools/file/toolkit.py (hunk 1 above)
Comment (ruanrongman) [reply promptulate/tools/file/toolkit.py:36]: use simpler definition ,like 
```python
file_tools = FileToolKit(root_dir="./xxx",modes="rwa")
agent = ToolAgent(tools=file_tools)
```

File: promptulate/tools/file/toolkit.py (hunk 5 above)
Comment (ruanrongman) [reply promptulate/tools/file/toolkit.py:47]: [Undertone0809](https://github.com/Undertone0809) implement is better.

File: promptulate/tools/file/tools.py (hunk 7 above)
Comment (ruanrongman) [reply promptulate/tools/file/tools.py:63]: use funtion implement , that agent can konw the args.

**Review by ruanrongman — APPROVED** [review]

File: tests/tools/test_file_tools.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        assert toolkit.root_dir == os.getcwd()
+
+        # 测试当root_dir为空字符串时，root_dir会被赋值为当前工作目录
+        toolkit1 = FileToolKit(root_dir="")
+        assert toolkit1.root_dir == os.getcwd()
+
+        # 测试当root_dir为有效路径时，root_dir会被赋值为该路径
+        toolkit2 = FileToolKit(root_dir=self.temp_dir)
+        assert toolkit2.root_dir == self.temp_dir
+
```
[the remark this answers is not in the document]
Comment (ruanrongman) [reply tests/tools/test_file_tools.py:34]: English is better

File: tests/tools/test_file_tools.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        # 测试当selected_tools包含存在的工具时，get_tools()返回对应的工具
+        toolkit2 = FileToolKit(root_dir=self.temp_dir, modes=["write", "read"])
+        tools = toolkit2.get_tools()
+        assert len(tools) == 2
+
+        # 测试当selected_tools包含不存在的工具时，会抛出ValueError
+        with pytest.raises(ValueError):
+            FileToolKit(root_dir=self.temp_dir, modes=["write", "nonexistent"])
+
```
[the remark this answers is not in the document]
Comment (ruanrongman) [reply tests/tools/test_file_tools.py:66]: English is better

File: tests/tools/test_file_tools.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        assert os.path.exists(f"{self.dir_name2}/{self.temp_file}")
+
+    def test_move_file_tool(self):
+        # 测试MoveFileTool的run方法
+        with open(f"{self.dir_name}/{self.temp_file}", "w") as file:
+            file.write("test")
+        Mtools = MoveFileTool(self.dir_name)
+        Mtools.run(self.temp_file, self.dir_name2)
+        assert os.path.exists(f"{self.dir_name2}/{self.temp_file}")
+        assert not os.path.exists(f"{self.dir_name}/{self.temp_file}")
```
[the remark this answers is not in the document]
Comment (ruanrongman) [reply tests/tools/test_file_tools.py:125]: English is better

File: tests/tools/test_file_tools.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ruanrongman) [reply tests/tools/test_file_tools.py:66]: use model = "rwa" is better

**Review by ruanrongman — CHANGES_REQUESTED** [review]

**Comment by Undertone0809** [comment]
image

@ruanrongman It's a good solution, but how to handle `copy, list, delete ...` operations?

**Comment by Undertone0809** [comment]
image

@ruanrongman He can add Parameters to show args. For pne's tool, maybe class type tool is better. For custom tool, it's more simple for developer.

File: promptulate/tools/file/tools.py (hunk 7 above)
Comment (Undertone0809) [reply promptulate/tools/file/tools.py:63]: @ruanrongman He can add Parameters to show args. For pne's tool, maybe class type tool is better. For custom tool, it's more simple for developer.

**Comment by longsihua2003** [comment]
In this case, how do I deal with the situation that I have multiple tools to use，And it seems hard to understand。
截屏2024-03-04 19 52 12
@ruanrongman

File: promptulate/tools/file/tools.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+                "Unicode decode error:"
+                "The file could not be decoded with utf-8 encoding."
+            )
+        except IOError:
+            return "An error occurred while trying to read the file."
+
+
+class DeleteFileTool(Tool):
+    name: str = "delete-file"
+    description: str = (
```
Comment (Undertone0809) [inline promptulate/tools/file/tools.py:118]: All tool should add parameters, parameters description should not declare in `description`. It means you should remove `"The output parameters must be: file_name(file name)."` 

See https://github.com/Undertone0809/promptulate/blob/d0af1e59d6f67b9f35a3cab560fac9a1ccd57442/tests/tools/test_tool.py#L172 to see how to declare a tool.

**Review by Undertone0809 — CHANGES_REQUESTED** [review]

**Comment by ruanrongman** [comment]
> image
> 
> @ruanrongman It's a good solution, but how to handle `copy, list, delete ...` operations?

can reference the linux,define the code

**Comment by ruanrongman** [comment]
> r means read, w means write, a means append. Like os.open function argument.

yes

File: promptulate/tools/file/tools.py (hunk 7 above)
Comment (ruanrongman) [reply promptulate/tools/file/tools.py]: > @ruanrongman He can add Parameters to show args. For pne's tool, maybe class type tool is better. For custom tool, it's more simple for developer.

ok,class type is better

**Review by ruanrongman — CHANGES_REQUESTED** [review]
I think you can reference the linux, Like os.open function argument.

**Comment by Undertone0809** [comment]
> > image
> > @ruanrongman It's a good solution, but how to handle `copy, list, delete ...` operations?
> 
> can reference the linux,define the code

Feel bad, Linux pattern lists directories with `os.listdir()`, and using `os.remove()` to delete file. `rwa` can only control the mode  of file operations, but it is not suitable for directory operations.

**Review by Undertone0809 — APPROVED** [review]
Add parameters attribution for all tools plz.

File: promptulate/tools/file/tools.py [hunk 14]
```diff
@@ -2,6 +2,7 @@
 import shutil
 
 from promptulate.tools.base import Tool
+from pydantic import Field
```
Comment (Undertone0809) [inline promptulate/tools/file/tools.py:5]: use `from promptulate.pydantic_v1 import BaseModel, Field` to import pydantic. In order to compatible pydantic v1 and pydantic v2.

ref: https://github.com/Undertone0809/promptulate/pull/483

**Review by Undertone0809 — CHANGES_REQUESTED** [review]

**Comment by Undertone0809** [comment]
Add parameters attribution for all tools plz. See the following demo to define a tool.

image

File: promptulate/tools/file/tools.py [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+        "Append the content to the end of the file."
+        "If the file does not exist, edit it after it is created"
+    )
+    parameters = ToolParameters
+
+    def __init__(self, root_dir: str, *args, **kwargs) -> None:
+        self.root_dir = root_dir
+        super().__init__(*args, **kwargs)
+
+    def _run(self, file_name: str, text: Optional[str]) -> str:
```
Comment (Undertone0809) [inline promptulate/tools/file/tools.py:76]: Optional means your default value is None

File: promptulate/tools/file/tools.py [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+
+class RToolParameters(BaseModel):
+    file_name: str = Field(
+        description="The name of the file you want to edit.",
+    )
+
+
+class ToolParameters(RToolParameters):
+    text: Optional[str] = Field(
```
Comment (Undertone0809) [inline promptulate/tools/file/tools.py:16]: Optional means your default value is None

File: tests/tools/test_file_tools.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+        # root_dir will be assigned to the given directory
+        toolkit2 = FileToolKit(root_dir=self.temp_dir)
+        assert toolkit2.root_dir == self.temp_dir
+
+    def test_validate_tools(self):
+        # Test when mode_List is None,
+        # tools will be assigned to all tools
+        toolkit = FileToolKit(root_dir=self.temp_dir, modes=None)
+        tools = toolkit.get_tools()
+        assert len(tools) == 7
```
Comment (Undertone0809) [inline tests/tools/test_file_tools.py:43]: The detail, you need to assert whether relevant tool is in `tools`

File: tests/tools/test_file_tools.py [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+    # Create a second temporary directory
+    temp_dir2 = tempfile.TemporaryDirectory()
+    dir_name2 = temp_dir2.name
+
+    # Create a file
+    temp_file = "test.txt"
+
+    def test_list_directory_tool(self):
+        # Test ListDirectoryTool's run method
+        Ltools = ListDirectoryTool(self.dir_name)
```
Comment (Undertone0809) [inline tests/tools/test_file_tools.py:81]: plz naming conventions using lowercase underline

Comment (Undertone0809) [inline tests/tools/test_file_tools.py:81]: You'd better assert Ltools.to_schema() == {xxx}. The following tools are the same.

**Review by Undertone0809 — CHANGES_REQUESTED** [review]

**Comment by Undertone0809** [comment]
lgtm

**Review by ruanrongman — APPROVED** [review]

**Review by Undertone0809 — APPROVED** [review]

**Pull request merged**